### DISCUSSION TITLE
Questions regarding the future of GRDB + SQLCipher

### DISCUSSION THREAD

**Discussion by groue** [groue/GRDB.swift#1495] [Q&A]
Hello all,

[GRDB 7, presently in the making](https://hachyderm.io/@groue/111907921298623607), will drop CocoaPods support. 💔 

Some folks will be able to migrate to SPM. But folks who use **GRDB + SQLCipher will be stuck with GRDB 6**.

This sad situation is a conjunction of several blockers:

- [SPM can't build SQLCipher](https://github.com/sqlcipher/sqlcipher/issues/371).
- [CocoaPods won't publish a new GRDB version from Xcode 15.3](https://github.com/CocoaPods/CocoaPods/issues/11839) (GRDB 7 will require Swift 5.10+ and Xcode 15.3+).
- I'd rather not ship binary builds of SQLCipher, because there are always people who want a different build (different compiler version, different platform, whatever). Binary builds are a chore and a time sink, but even more so for a library that aims at running on whatever device can run Swift and SQLite. And, well, I have to admit I'm not a build expert.

I will not remove the SQLCipher and CocoaPods support from the repository, because the CocoaPods blocker might be resolved one day.

---

I have a question for the community: do you have any advice or experience that could help unblocking the situation and allow SQLCipher users to enjoy the upcoming GRDB 7?

Also, if you are using SQLCipher and are concerned with this announcement, did you consider using [data protection](https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/encrypting_your_app_s_files#) instead?

Please share your experiences below 🙏 

[1]: https://github.com/sqlcipher/sqlcipher/issues/371
[2]: https://github.com/CocoaPods/CocoaPods/issues/11839

**Comment by groue** [comment]
Hello @ayoy. May I ask you, one of the authors of https://github.com/duckduckgo/GRDB.swift (the GRDB fork that ships a pre-built XCFramework of GRDB + SQLCipher), if you can imagine anything that could be done from here, the main repository?

**Reply by ayoy** [reply]
Hi @groue! Thanks for the ping and congratulations on dropping CocoaPods :)

At DuckDuckGo we've initially used GRDB with SQLCipher in SPM as source code, and have switched to binary framework only to speed up our compile times. But my point is that it's perfectly doable to have it distributed via SPM as source code.

That said, it will definitely be tricky to make SQLCipher-enabled GRDB available alongside vanilla GRDB in the same SPM package (code repository). And even if it's doable, it would still require you to include SQLCipher sources in your repository, and it will still force users to use the fixed set of compile flags for SQLCipher.

You'll find our source distribution in the [SQLCipher](https://github.com/duckduckgo/GRDB.swift/blob/SQLCipher/Package.swift) branch of our fork (we keep it as a backup in case binary frameworks stop working). If you're not so keen on digging into the build system, I'd be happy to try it myself (i.e. to see if it's possible to have GRDB and GRDB+SQLCipher targets available in a single SPM package). If that works for you, just let me know when you have the GRDB 7 branch ready to fork and play with.

Thanks for your great work! 💪

**Reply by groue** [reply]
Right now I'm focusing on the features (with a lot of boxing with Swift Concurrency 🥊😜), and probably I'll release the lib without SQLCipher support initially. And we also need to give time to people who want to comment, contribute, or mourn here. But I'll let you know, yes, because you know a heck a lot more than I do!

**Comment by groue** [comment]
Hello @michaelkirk and @charlesmchen. We had conversation about GRDB in the Signal app in the past. [Signal currently uses GRDB+SQLCipher through CocoaPods](https://github.com/signalapp/Signal-iOS/blob/e6ba17aa451b6fa66ccc492b078a0d5d41e6e8d6/Podfile#L27), so I thought you'd be happy to know about the situation. I'm curious about your opinions.

**Reply by groue** [reply]
cc @max-signal, @harry-signal, @sashaweiss-signal, @pete-signal, because you are the main contributors to Signal in the last year.

**Comment by bok-** [comment]
We use GRDB + SQLCipher with a small script that wraps CocoaPods to generate XCFrameworks, which I'd love to do away with. Unfortunately I don't have any good ideas off the top of my head as to how you could accomplish that with SQLCipher not supporting SPM themselves, and assuming you had that - without forcing it to be a dependency for everyone regardless of whether they want it or not, I guess for the latter you could create a GRDB-SQLCipher wrapper package that depends on both, but without a way to pass configuration , compiler directives or anything really between packages it feels pretty difficult. If you do include it in the main package you can probably usage the `package` access modifier to allow the main package to expose things to allow a `GRDB+SQLCipher` target to overwrite it, but that feels like it could be a large refactor.

We do also use [data protection](https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/encrypting_your_app_s_files#) but its not something you can rely on in a number of situations — such as when the user disables their iPhone's passcode. Yes the user is compromising their own information security so its their own fault, but we would likely still suffer reputation damage if that happened (I work for a large bank), so we want to make sure we have full control over the encryption of the database. Hope that context helps some.

**Reply by groue** [reply]
> We do also use [data protection](https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/encrypting_your_app_s_files#) but its not something you can rely on in a number of situations — such as when the user disables their iPhone's passcode. Yes the user is compromising their own information security so its their own fault, but we would likely still suffer reputation damage if that happened (I work for a large bank), so we want to make sure we have full control over the encryption of the database. Hope that context helps some.

Thanks @bok- :-) That's the context I was missing.

Maybe you can answer two other questions (I don't want to infringe any industrial property):

- If you need to change the key, are you satisfied with the [recommended technique](https://github.com/groue/GRDB.swift/blob/master/README.md#changing-the-passphrase-of-an-encrypted-database), or did you have to do something differently?
- How do you build the key? (now I may cross some limits, so please don't answer if you have any reasonable doubt 😅)

**Reply by bok-** [reply]
Apologies for the late reply! I personally think it's not strictly industrial property when you take a standard approach, so I don't think it's infringing anything to describe what we do.

I decided early on that changing the paraphrase on the database was an expensive operation so we wanted to avoid that. As a result we always encrypt the database using a symmetric key generated by the best option CryptoKit gives you out of the box. We had some performance problems with this initially but found it was because we supplied the key to GRDB in the wrong format. Since we fixed that it has been perfect.

That key is stored in keychain with flags for no backup + no migration. So during normal operation that's accessible to the app for synchronisation or user access or whatever.

If we decide that a user's device is not safe because they don't have a passcode set, or it's jailbroken or other things OWASP says to check for (and some of our own custom stuff) then we take the symmetric key from the keychain, and re-encrypt it with a key derived from the PIN the user uses to access our app. That also means any situation where the PIN was previously optional it is now required. It also means we don't need to change the paraphrase protecting our GRDB database, just adding an extra step to decrypt the key we pass to GRDB.

**Reply by groue** [reply]
This is a great answer, thank you very much 🙏

**Comment by groue** [comment]
**It should be possible for applications to keep on using GRDB+SQLCipher with CocoaPods** even if it is [impossible](https://github.com/CocoaPods/CocoaPods/issues/11839) to publish a new version of the GRDB pod, by declaring a dependency on a branch:

```ruby
# Same as pod 'GRDB.swift/SQLCipher', '~> 6.0'
# We depend on a branch due to https://github.com/CocoaPods/CocoaPods/issues/11839
# that prevents GRDB from publishing tagged releases on cocoapods.org.
pod 'GRDB.swift/SQLCipher', git: 'https://github.com/groue/GRDB.swift.git', branch: 'GRDB6'
pod 'SQLCipher', '~> 4.0'
```

The GRDB6 will only ever point to tagged releases. For example, on 2024/03/24, GRDB6 is identical to v6.26.0.

The same technique will be available for GRDB7 (on the GRDB7 branch - which is currently a work in progress and is NOT production-ready).

**Comment by rakuyoMo** [comment]
Hello! I have a question: When I rely on a sqlcipher fork that supports spm through spm (such as https://github.com/Tencent/sqlcipher/blob/v1.4.4/Package.swift), can GRDB's encryption function be used normally?

(Current version, and upcoming GRDB7)

**Reply by groue** [reply]
Who knows? Try and share your experience.

**Reply by rakuyoMo** [reply]
I want to say I failed

I use Tuist+SPM, which is equivalent to directly operating multiple xcprojects to some extent.

It may work. I modify the settings of GRDB to depend on a SQLCipher binary package, but it is just a component, so I need to modify every project that depends on this component.

Maybe using the above-mentioned https://github.com/duckduckgo/GRDB.swift or other forks containing SQLCipher binary packages is a better choice.

**Reply by rakuyoMo** [reply]
There are so many forks of GRDB (I guess because it's good enough), so many that I only found [duckduckgo/GRDB.swift](https://github.com/duckduckgo/GRDB.swift). If anyone knows of other good GRDB forks that include SQLCipher, please share them.

**Reply by groue** [reply]
The [Network Graph](https://github.com/groue/GRDB.swift/network) of the repository can be useful when you are looking at specific forks. You can see individual commits, and their purpose (if they are well written). It's still the search of a needle in a haystack, but with a magnifying glass ;-)

**Comment by rakuyoMo** [comment]
Have you ever considered separating the SQLCipher-related code from GRDB's main repository (`Package.swift`) and putting it in an independent repository?

As mentioned above, I tried to rely on the SQLCipher binary library, but there were many changes to be made. And it would be very troublesome to synchronize GRDB upstream updates in the future.

If the SQLCipher-related functions are separated from GRDB, then users including me can rely on GRDB and the SQLCipher binary library to make their own component library, which is more flexible. And the amount of SQLCipher-related function code is not large, which is convenient to rely on in the form of source code (or other methods) in this secondary development component. Unless there are special circumstances, users do not need to care about GRDB updates, and can use the latest version of GRDB in the project without updating this secondary development component.

**Reply by rakuyoMo** [reply]
Let me be more specific. Suppose there are two spm libraries, `GRDB` and `GRDB-SQLCipher`. `GRDB` is all the code in GRDB.swift except for the code related to SQLCipher. `GRDB-SQLCipher` contains the code related to SQLCipher.

Users can create a new component, such as `MyGRDB`, which depends on `GRDB` and a SQLCipher binary package, and also adds the content in `GRDB-SQLCipher` in the form of source code.

If `GRDB` is updated in the future (for example, to version 7.1), users do not need to update `MyGRDB` (assuming that `GRDB` 7.1 will not affect `GRDB-SQLCipher`), and can directly use `GRDB` 7.1 in the main project.

**Reply by rakuyoMo** [reply]
Otherwise, as it is now, I need to fork GRDB.swift and add a dependency on SQLCipher in Package.swift (and I may need to modify some code, I'm not sure).

Then I need to add some version control, add a tag or create a branch. Finally, use the tag or branch to reference it in the project.

If GRDB is updated later, I need to pay some cost to synchronize the upstream update, then create a new tag or branch, and finally reference the new version in my project.

This process is not as simple and once and for all as the above method for me.

Of course, you or other users may not think so, which is normal. These words only represent my thoughts. We are just discussing, discussing a possibility, right?

**Reply by groue** [reply]
No, that's not a possibility.

An hypothetical `GRDB-SQLCipher` package can not instruct its dependency `GRDB` to stop linking against the system SQLite, and link against SQLCipher instead. That's just not an SPM feature.

**Reply by groue** [reply]
Thanks for searching, though :-) But, yeah, we have technical (and time) constraints to deal with.

**Reply by ThePragmaticArt** [reply]
Wait, couldn't the library be separated into 3 layers? 

At a Foundation layer, you would have the DI interfaces to interface with SQLCipher/SQLite shared interfaces if necessary. At a Domain layer the core GRDB interfaces. Optionally, there can be a separate interface with default SQLite definitions for injection purposes which the consuming project would be responsible for importing -> Injecting into GRDB itself?

This would at least make GRDB agnostic of binary implementations and hopefully provide injection points to override some behaviors via SQLCipher if desired.

If someone would like to use vanilla SQLIte, there's just an initial definition that would be required to be DI'ed into the library. Having a non 1:1 coupling between the core GRDB interfaces and core SQLite interfaces should leave the responsibility on the integration libraries to translate things correctly (given an opportunity) where-as GRDB can focus on the vanilla SQLite translations within their own library exclusively.

Ideally the concept wouldn't be very far off from normal DI mechanisms or plugin approaches: https://github.com/apple/swift-log. 

The biggest headache today is that Realm is essentially dead (in a sense) and all 3rd party encryption libraries for CoreData are dead (given how long that's been a thing it's unlikely to be revived). The only remaining option for enterprise is SQLite encryption or plain encrypted files on disk...

**Comment by morajabi** [comment]
@rakuyoMo We've updated DuckDuckGo's awesome fork to use latest GRDB 7.0.0-beta.6, feel free to use! 

https://github.com/inline-chat/GRDB.swift

**Reply by morajabi** [reply]
Update: had to patch GRDBQuery to work with it, it's here: https://github.com/inline-chat/GRDBQuery

**Reply by morajabi** [reply]
As we open sourced our client Swift apps, there is a (simple) example of using GRDB + SQLCipher in a production application: https://github.com/inline-chat/inline/blob/main/apple/InlineKit/Sources/InlineKit/Database.swift#L176 

Hope it helps someone!

**Reply by ChaosCoder** [reply]
We have not found a fork that worked with the release version of GRDB 7, so we pushed a new fork at [TICESoftware/GRDB-SQLCipher](https://github.com/TICESoftware/GRDB-SQLCipher) with GRDB 7.3.0 + SQLCipher 4.6.1.

I would love to see this being maintained by some official party, but for now, we will maintain this repository – issues and PRs are welcome. 

Related: https://github.com/sqlcipher/sqlcipher/issues/371

**Reply by morajabi** [reply]
@ChaosCoder what issue did you encounter with the inline-chat/GRDB.swift fork? I'd be happy to fix because we intend to maintain it. Is it just the version? I'll update it to use the release version.

**Reply by aehlke** [reply]
@ChaosCoder @morajabi is it documented / specified somewhere in your repos which xcode version is used for the xcframework? just want to understand how to determine and how to update it myself, thanks

**Reply by ChaosCoder** [reply]
@morajabi Those points are what we fixed with our fork:
- Your fork was out of date and did not have any of the releases from the last months. Especially there was no release version of GRDB 7.
- The xcframework was compiled with a version of Xcode/Swift, that had the following bug: https://github.com/groue/GRDB.swift/issues/1596 and could therefor not be used
- The script did not produce a xcframework out of the box for the new GRDB releases
- You don't have the issues-tab activated in GitHub, so I could not open an issue with what I found

If you want to maintain your fork, maybe you could try to automate the release process so it is tied to the release of new GRDB and/or SQLCipher versions?

@aehlke I used Xcode 16.2, but you are right, it was not specified. I just updated the readme so it includes this information. The script itself just uses the Xcode that was selected via `xcode-select` on your machine and now includes this information automatically. See https://github.com/TICESoftware/GRDB-SQLCipher/commit/e374dc6524b3fee8bc5a32f8fefb2ffc7592f63a

**Reply by aehlke** [reply]
Thanks so much. I'm not using SQLCipher but a lot of my build time is taken up by GRDB so I'd like to use xcframeworks. Appreciate it

**Reply by ayoy** [reply]
Just a note for anyone relying on the DuckDuckGo fork – I’ve finally [updated it to GRDB 7.4.1](https://github.com/duckduckgo/GRDB.swift/releases/tag/3.0.0). Sorry for the delay here, but we just didn’t need to upgrade earlier.

**Reply by morajabi** [reply]
@ChaosCoder Thanks for the detailed comment. I fixed all of the issues you mentioned and released a new version. 

If anyone uses GRDBQuery, I think they'd have to also use [the patched fork](https://github.com/inline-chat/GRDBQuery) that points to the forked GRDB. 

Also, as a side note, It's amazing to see the active community around GRDB. Special thanks to @ayoy for creating the initial fork.

**Comment by DrewKiino** [comment]
would love to see a native SPM approach to this as we will be integrating GRDB in an enterprise app and we will be relying on encryption later down the line - fingers crossed

**Reply by aehlke** [reply]
use a fork https://github.com/groue/GRDB.swift/discussions/1495#discussioncomment-12674158

**Comment by c128128** [comment]
Hey guys, check this pls https://github.com/sqlcipher/sqlcipher/issues/371#issuecomment-3156623200

**Comment by mezhevikin** [comment]
Hello everyone. I created a fork of GRDB in which the official SQLCipher.swift package is enabled by default.

https://github.com/mezhevikin/GRDB.SQLCipher.swift

I am looking forward to official SQLCipher support, but for now I will use this temporary solution — maybe it will be useful to someone.

Only two files were changed:
- **Package.swift** – added the SQLCipher.swift dependency and enabled the `GRDBCIPHER` and `SQLITE_HAS_CODEC` flags.
- **Database.swift** – added a direct import of SQLCipher, otherwise the compiler could not find the `sqlite3_key` function.

All other files remain unchanged.

All tests passed except for `testSQLCipher3Compatibility`.

**Reply by groue** [reply]
I advise not to work from the current `development` branch, which improperly defines package traits and is not fixed yet (see https://github.com/groue/GRDB.swift/pull/1827#issuecomment-3418155348 for the problems it creates).

If you're working from the current `master` branch (the latest stable release), adding two `swiftSettings` in your forked Package.swift should be sufficient:

```swift
.define("SQLITE_HAS_CODEC"),
.define("GRDBCIPHER"),
```

**Reply by shivpatel94** [reply]
Hi guys, I've been following this thread. 

First off, a huge amount of gratitude to you @groue for all you do for us! GRDB is unbeatable.

I'm wondering if the plan is to maintain an official GRDBCipher fork or to support users with their fork in the long run? Obviously I have a strong preference on the official fork route but can understand that it creates a lot of unnecessary work on top of all the work you already do for us! 

Thank you.

**Reply by groue** [reply]
Hello @shivpatel94. The latest development occurred in what some people wittily call the "GRDB SQLCipher saga" :-) #1827.

My latest opinion is [there](https://github.com/groue/GRDB.swift/pull/1827#issuecomment-3450562202).

To make a very quick summary: I'll first make it easier to fork GRDB. I'll then consider maintaining a fork. I'm very slow at that because this is a nasty energy sucker — zero fun, many disappointments.

**Comment by groue** [comment]
GRDB v7.10.0 has shipped, with an easier way to use GRDB + SQLCipher with SPM. Please check https://github.com/groue/GRDB.swift/blob/master/README.md#encryption

**Reply by morajabi** [reply]
@groue thank you so much 🥹

**Reply by shivpatel94** [reply]
@groue awesome! Thank you so much.

As always, hugely grateful for all the work you have put and continue to put into GRDB. I hope you hear that enough!

Truly standing on the shoulder of giants

**Reply by groue** [reply]
Thank you! And thank you to @R4N and @marcprux, whose generosity has been key on this release!

**Reply by morajabi** [reply]
FWIW I updated our forks to match the official instructions: 
https://github.com/inline-chat/GRDB.swift
https://github.com/inline-chat/GRDBQuery

**Reply by Alex-Hoth** [reply]
Just wanted to note that today's beta release of Xcode 26.4 beta 1 "Added the ability to enable package traits on dependencies from the Package Dependencies view. (141748785)"

**Reply by groue** [reply]
Great! But don't hold your breath! 🧵 https://hachyderm.io/@groue/116082053586557129

**Comment by mezhevikin** [comment]
@groue Good news! Is it worth waiting for an official fork with SQLCipher included under the groue or sqlcipher account?

Or does everyone have to create their own fork and keep it up to date?

**Reply by groue** [reply]
Hello @mezhevikin! Waiting is not a good strategy here.

**Reply by groue** [reply]
I'm not _happy_ that forking is necessary. But that's the only available option I can provide right now.

**Reply by mezhevikin** [reply]
@R4N
It's fine if you don't want to fork your account, but maybe the guys at SQLChiper are willing to do that? I'd prefer to have a single entry point instead of a million forks.

**Reply by groue** [reply]
Your preference and the preferences of all GRDB and SQLCipher users are well considered, by both me and people at SQLCipher. Please do not apply inconsiderate pressure.

**Reply by R4N** [reply]
Ongoing discussion about a standardized fork is opened over here: https://github.com/groue/GRDB.swift/discussions/1854
